docs/proposals/generalize-modules.md

```diff
@@ -0,0 +1,72 @@
+---
+title: "Generalize Modules Service to make it extensible"
+linkTitle: "Generalize Modules Service to make it extensible"
+weight: 1
+slug: generalize-modules
+---
+
+- Author: @annanay25
+- Reviewers: 
+- Date: April 2020
+- Status: Draft
+
+## Overview
+
+Cortex uses modules to start and operate services with dependencies. Inter-service dependencies are specified in a map and passed to a module manager which ensures that they are started in the right order of dependencies. While this works really well, the implementation is tied in specifically to the Cortex struct and is not flexible for use with other projects like Loki, which also require similar forms of dependency management. 
+ 
+We would like to extend modules in cortex to a generic dependency management framework, that can be used by any project with no ties to cortex. 
+
+## Specific goals
+
+- Framework should allow for reusing cortex modules and allow us to:
+  - Add new modules
+  - Overwrite the implementation of a current module
+  - Manage dependencies
+- Framework should allow for building an application from scratch using the `modules` package, with no dependencies on Cortex. For ex: Remove code from Loki that was copied from `pkg/cortex/cortex.go`.
+
+
+
+## Proposed Design
+
+To make the modules package extensible, we need to abstract away any Cortex specific details from the module manager. The proposed design is to -
+
+- Make a new component `moduleManager`, which is envisioned to be a central manager for all modules of the application. It stores modules & dependencies, and will be housed under a new package `pkg/util/modules`. The following is the interface for interacting with the `moduleManager`:
+```
+// Manager defines the interface for interaction with moduleManager.
+type Manager interface {
+   RegisterModule(name string, deps []string, svc service, options ...Option)
+   AddDependency(fromModule string, toModule string) error
+   ModuleServiceWrapper(name string, modServ services.Service)
+   InitModuleServices(target string) (map[string]services.Service, error)
+}
+```
+
+- Modules can be created by the application and registered with the `ModuleManager` using `RegisterModule`. The parameters are:
+  - `name`: Name of the module
+  - `deps`: A list of modules that this module depends on to run.
+  - `svc`: A function that will be used to start the module.
+  - `options`: Options are a variadic function parameter that can be used by the `moduleManager` to wrap the service function. In Cortex, we pass `Manager.ModuleServiceWrapper`, which wraps service to work with dependencies. 
```

Comment:
The reason why `Server` service is special in Cortex is because underlying "Server" from weaveworks package handles SIGINT/SIGTERM, and reacts on it by exiting from its `Run` method, while keeping HTTP server running. This makes `Server` service transition to Stopping state, which is a signal for Cortex to begin shutdown procedure.

Notice that Server service still waits for other services to stop first (!) before stopping HTTP server.

If "Server" service was wrapped in "module wrapper", Cortex would only see outer (wrapping) service -- which waits until underlying service is fully stopped before moving to Stopping state.

If we 1) disable weaveworks' Server handling of SIGTERM/SIGHUP (not currently possible), and 2) make Cortex wait for SIGHUP/SIGTERM on it's own, then we can also use module wrapper for Server.

Reply:
Thanks for the detailed explanation @pstibrany! This design doc addresses this problem by not wrapping the `Server` service in module wrapper (by passing `nil` options) like so - 
```
mm := modules.NewManager()
.
.
mm.RegisterModule(Server, nil, t.initServer)
```

.. and by wrapping other services in module wrapper like so -

```
mm.RegisterModule(Ring, []string{API, RuntimeConfig, MemberlistKV}, t.initRing, mm.ModuleServiceWrapper)
```

Reply:
I'm not sure we even need `Options`. What other options do you expect to have? We can always introduce variadic options in the future, when needed.

Reply:
Hrm, I'm not sure but this could enable extra-wrapping of services, as you mentioned in the other comment https://github.com/cortexproject/cortex/pull/2542/files#r417929921

Reply:
I think stuff that can be done by caller should be done by caller. When we see what other options are interesting, we can always introduce them. But I would start with simpler design (unless we already have some options now)